fastcrypto-zkp/src/bn254/poseidon/mod.rs

```diff
@@ -1,65 +1,82 @@
 
-impl Debug for PoseidonWrapper {
-    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
-        f.debug_struct("PoseidonWrapper").finish()
-    }
+        // Neptune returns the state element with index 1 but we want the first element.
+        poseidon.elements[0]
+    }};
 }
 
-impl Default for PoseidonWrapper {
-    fn default() -> Self {
-        Self::new()
+/// Poseidon hash function over BN254.
+pub fn hash(inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
+    if inputs.is_empty() || inputs.len() > 16 {
+        return Err(FastCryptoError::InputLengthWrong(inputs.len()));
     }
-}
 
-impl PoseidonWrapper {
-    /// Initialize a Poseidon hash function.
-    pub fn new() -> Self {
-        Self {
-            instance: Poseidon::new(),
-        }
-    }
-
-    /// Calculate the hash of the given inputs.
-    pub fn hash(&self, inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
-        self.instance
-            .hash(inputs)
-            .map_err(|_| FastCryptoError::InvalidInput)
-    }
+    // Instances of Poseidon and PoseidonConstants from neptune have different types depending on
+    // the number of inputs, so unfortunately we need to use a macro here.
+    let result = match inputs.len() {
+        1 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U1),
+        2 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U2),
+        3 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U3),
+        4 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U4),
+        5 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U5),
+        6 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U6),
+        7 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U7),
+        8 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U8),
+        9 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U9),
+        10 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U10),
+        11 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U11),
+        12 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U12),
+        13 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U13),
+        14 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U14),
+        15 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U15),
+        16 => define_poseidon_hash!(inputs, POSEIDON_CONSTANTS_U16),
+        _ => return Err(InvalidInput),
+    };
+    Ok(fr_to_bn254fr(result))
 }
 
 /// Calculate the poseidon hash of the field element inputs. If the input length is <= 16, calculate
 /// H(inputs), if it is <= 32, calculate H(H(inputs[0..16]), H(inputs[16..])), otherwise return an
 /// error.
 pub fn to_poseidon_hash(inputs: Vec<Fr>) -> Result<Fr, FastCryptoError> {
-    static POSEIDON: OnceCell<PoseidonWrapper> = OnceCell::new();
-    let poseidon_ref = POSEIDON.get_or_init(PoseidonWrapper::new);
     if inputs.len() <= 16 {
-        poseidon_ref.hash(inputs)
+        hash(inputs)
     } else if inputs.len() <= 32 {
```

Comment:
So 32 is the max? and if someone has a longer input, they should split beforehands and recursively invoke that function? Do we want to offer that recursive (for loop) function, to accommodate longer inputs as well (where internally it does the splits and merge for any size of inputs?

Reply:
We could definitely do that. This design is aligned with the TS code due to zklogin. I think we should keep it as is for now and then document it more clearly.

Reply:
@jonas-lj can we at least document that this is the case and we plan to update in the future? If used generically (ie in smart contracts) they should know the limits.

Reply:
Sure. Now that I think more about it, we could also extend it to more inputs now by using a Merkle tree with degree 16. Then it'll me aligned with the circuit for <= 32 inputs but still allow for more.